File: backend/src/agents/security.js
import { invokeClaude } from "../bedrock/bedrockClient.js";
import {
  getSecurityHubFindings,
  getGuardDutyFindings,
} from "../tools/securityTools.js";

const SYSTEM = `You are the Security agent.
You review SecurityHub and GuardDuty findings and surface the most important risks.
Prioritise: public S3 buckets, IAM wildcard policies, open security groups, GuardDuty alerts.
For each finding give a one-line remediation. Be direct about severity.`;

export async function runSecurity() {
  const [sh, gd] = await Promise.all([
    getSecurityHubFindings({ max: 25 }),
    getGuardDutyFindings({ max: 25 }),
  ]);

  const findings = [
    ...(Array.isArray(sh) ? sh : []),
    ...(Array.isArray(gd) ? gd : []),
  ];

  const res = await invokeClaude({
    system: SYSTEM,
    messages: [
      {
        role: "user",
        content: `Findings: ${JSON.stringify(
          findings.slice(0, 20)
        )}. Summarise the critical ones and give remediation steps.`,
      },
    ],
    maxTokens: 800,
  });

  const counts = findings.reduce(
    (a, f) => {
      a.total++;
      if (f.severity === "crit") a.critical++;
      else if (f.severity === "warn") a.medium++;
      else a.low++;
      return a;
    },
    { total: 0, critical: 0, medium: 0, low: 0 }
  );

  return {
    summary: res.content?.find((b) => b.type === "text")?.text || "",
    findings,
    counts,
  };
}

File: backend/src/agents/security.py
from ..bedrock.bedrock_client import run_agent_loop
from ..tools.security_tools import get_guard_duty_findings, get_security_hub_findings

SYSTEM = """You are the Security agent.
Call the tools to fetch SecurityHub and GuardDuty findings, then surface the most
important risks. Prioritise: public S3 buckets, IAM wildcard policies, open security
groups, GuardDuty alerts. If the first batch looks incomplete or you need more context
on severity distribution, you may call a tool again with a higher max.
Present remediations as a "Suggestions:" section — one concrete, actionable line per
finding. Be direct about severity."""

TOOLS = [
    {
        "name": "get_security_hub_findings",
        "description": "Fetch active SecurityHub findings.",
        "input_schema": {
            "type": "object",
            "properties": {"max": {"type": "integer", "description": "Max findings to fetch, default 25"}},
        },
    },
    {
        "name": "get_guard_duty_findings",
        "description": "Fetch active GuardDuty findings.",
        "input_schema": {
            "type": "object",
            "properties": {"max": {"type": "integer", "description": "Max findings to fetch, default 25"}},
        },
    },
]


async def run_security():
    collected = []

    async def _tool_runner(name, input_):
        params = input_ or {}
        max_results = params.get("max", 25)
        if name == "get_security_hub_findings":
            res = await get_security_hub_findings(max=max_results)
        elif name == "get_guard_duty_findings":
            res = await get_guard_duty_findings(max=max_results)
        else:
            return {"error": "unknown tool"}
        if isinstance(res, list):
            collected.extend(res)
        return res

    result = await run_agent_loop(
        system=SYSTEM,
        user_message="Fetch SecurityHub and GuardDuty findings, then summarise the critical ones with remediation steps.",
        tools=TOOLS,
        tool_runner=_tool_runner,
    )

    counts = {"total": 0, "critical": 0, "medium": 0, "low": 0}
    for f in collected:
        counts["total"] += 1
        if f.get("severity") == "crit":
            counts["critical"] += 1
        elif f.get("severity") == "warn":
            counts["medium"] += 1
        else:
            counts["low"] += 1

    return {"summary": result["text"], "findings": collected, "counts": counts, "trace": result["trace"]}
